File: BSvsZP-Common -Java/Messages/ValidateTick.java
package Messages;

import org.omg.CORBA.portable.ApplicationException;
import Common.ByteList;
import Common.Tick;

public class ValidateTick extends Request {

    public short ComponentId;
    public Tick TickToValidate;
    private static int MinimumEncodingLength;

    protected ValidateTick(PossibleTypes type) {
        super(type);
    }

    public ValidateTick() {
        super(PossibleTypes.ValidateTick);
    }

    public ValidateTick(short componentId, Tick tick) {
        super(PossibleTypes.ValidateTick);
        ComponentId = componentId;
        TickToValidate = tick;
    }

    //new 
    public static ValidateTick Create(ByteList bytes) throws ApplicationException, Exception {
        ValidateTick result = null;

        if (bytes == null || bytes.getRemainingToRead() < ValidateTick.getMinimumEncodingLength()) {
            throw new ApplicationException("Invalid message byte array", null);
        } else if (bytes.PeekInt16() != (short) MESSAGE_CLASS_IDS.ValidateTick.getValue()) {
            throw new ApplicationException("Invalid message class id", null);
        } else {
            result = new ValidateTick();
            result.Decode(bytes);
        }

        return result;
    }

    @Override
    public void Encode(ByteList bytes) throws Exception {
        bytes.Add((short) MESSAGE_CLASS_IDS.ValidateTick.getValue());                              // Write out this class id first
        bytes.update();
        short lengthPos = bytes.getCurrentWritePosition();    // Get the current write position, so we
        // can write the length here later
        bytes.Add((short) 0);                             // Write out a place holder for the length
        bytes.update();

        super.Encode(bytes);                              // Encode the part of the object defined
        // by the base class
        bytes.AddObjects(ComponentId, TickToValidate);
        bytes.update();
        short length = (short) (bytes.getCurrentWritePosition() - lengthPos - 2);
        bytes.WriteInt16To(lengthPos, length);           // Write out the length of this object
        bytes.update();
    }

    @Override
    public void Decode(ByteList bytes) throws Exception {
        short objType = bytes.GetInt16();
        bytes.update();
        short objLength = bytes.GetInt16();
        bytes.update();

        bytes.SetNewReadLimit(objLength);
        bytes.update();
        super.Decode(bytes);
        ComponentId = bytes.GetInt16();
        bytes.update();
        TickToValidate = (Tick) bytes.GetDistributableObject();
        bytes.update();
        bytes.RestorePreviosReadLimit();
        bytes.update();
    }

    public short getComponentId() {
        return ComponentId;
    }

    public void setComponentId(short componentId) {
        ComponentId = componentId;
    }

    public Tick getTickToValidate() {
        return TickToValidate;
    }

    public void setTickToValidate(Tick tickToValidate) {
        TickToValidate = tickToValidate;
    }

    public static int getMinimumEncodingLength() {
        MinimumEncodingLength = 4 // Object header
                + 2 // ComponentId
                + 1;
        if (Common.ByteList.DEBUG)
        	System.out.println("ValidateTick.MinimumEncodingLength" + MinimumEncodingLength);
        return MinimumEncodingLength;
    }

    @Override
    public short getClassId() {
        return (short) MESSAGE_CLASS_IDS.ValidateTick.getValue();
    }

    @Override
    public int compareTo(Object o) {
        return 0;
    }

    @Override
    public MESSAGE_CLASS_IDS MessageTypeId() {
        return Message.MESSAGE_CLASS_IDS.ValidateTick;
    }
}

File: BSvsZP-Common/Messages/ValidateTick.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Common;

namespace Messages
{
    public class ValidateTick : Request
    {
        #region Private Properties
        private static Int16 ClassId { get { return (Int16)MESSAGE_CLASS_IDS.ValidateTick; } }
        #endregion

        #region Public Properties
        public override Message.MESSAGE_CLASS_IDS MessageTypeId() { return (Message.MESSAGE_CLASS_IDS)ClassId; }

        public Int16 ComponentId { get; set; }
        public Tick TickToValidate { get; set; }
        public static new int MinimumEncodingLength
        {
            get
            {
                return 4                // Object header
                       + 2              // ComponentId
                       + 1;
            }
        }
        #endregion

        #region Constructors and Factories

        /// <summary>
        /// Constructor used by factory methods, which is in turn used by the receiver of a message
        /// </summary>
        public ValidateTick() : base(PossibleTypes.ValidateTick) { }

        /// <summary>
        /// Constructor used by senders of a message
        /// </summary>
        /// <param name="username"></param>
        /// <param name="password"></param>
        public ValidateTick(Int16 componentId, Tick tick)
            : base(PossibleTypes.ValidateTick)
        {
            ComponentId = componentId;
            TickToValidate = tick;
        }

        /// <summary>
        /// Factor method to create a message from a byte list
        /// </summary>
        /// <param name="bytes">A byte list from which the message will be decoded</param>
        /// <returns>A new message of the right specialization</returns>
        new public static ValidateTick Create(ByteList bytes)
        {
            ValidateTick result = null;

            if (bytes == null || bytes.RemainingToRead < MinimumEncodingLength)
                throw new ApplicationException("Invalid message byte array");
            else if (bytes.PeekInt16() != ClassId)
                throw new ApplicationException("Invalid message class id");
            else
            {
                result = new ValidateTick();
                result.Decode(bytes);
            }

            return result;
        }

        #endregion

        #region Encoding and Decoding methods

        override public void Encode(ByteList bytes)
        {
            bytes.Add(ClassId);                              // Write out this class id first

            Int16 lengthPos = bytes.CurrentWritePosition;    // Get the current write position, so we
                                                                    // can write the length here later
            bytes.Add((Int16)0);                             // Write out a place holder for the length


            base.Encode(bytes);                              // Encode the part of the object defined
                                                                    // by the base class
            bytes.AddObjects(ComponentId, TickToValidate);

            Int16 length = Convert.ToInt16(bytes.CurrentWritePosition - lengthPos - 2);
            bytes.WriteInt16To(lengthPos, length);           // Write out the length of this object        
        }

        override public void Decode(ByteList bytes)
        {

            Int16 objType = bytes.GetInt16();
            Int16 objLength = bytes.GetInt16();

            bytes.SetNewReadLimit(objLength);

            base.Decode(bytes);

            ComponentId = bytes.GetInt16();
            TickToValidate = bytes.GetDistributableObject() as Tick;

            bytes.RestorePreviosReadLimit();
        }

        #endregion
    }
}
